docs/proposals/20231123-enhance-mid-tier-resource.md

```diff
@@ -0,0 +1,253 @@
+
+## Motivation
+
+This proposal introduces Mid+LS and Mid+BE to close the gap in Prod+LS and Batch+BE and fulfil the requirements of different task types.
+
+### Goals
+
+- How to calculate and update the mid-tier resource amount of a node.
+- Clarification of Mid-tier cgroup and QoS
+- Clarification of Mid-tier suppression and eviction policy
+- Necessary optimization of the scheduler/descheduler
+
+### Non-goals
+
+- Replace Batch-tier resources
+- Add new QoS type
+
+## Proposal
+
+### Terminology
+
+Here I would like to explain some concepts:
+1. koord-QoS
+
+Quality of Service, we assume that the same QoS level has similar operational performance, operational quality.
+
+2. koord-priority
+
+Scheduling priority, high priority can preempt low priority by default.
+
+3. Resource models
+
+We now have four resource models: prod, mid, batch and free. The resource model takes care of whether the resource is oversold and whether it is stable, which affects pod eviction.
+
+Note that koordinator bind koord-priority and the resource model; different priorities have different resource models.
+
+### User stories
+
+#### Story 1
+
+There are low priority, latency-sensitive service tasks whose performance requirements match those of Prod+LS, but which should not be suppressed but can tolerate being evictied as machine utilisation increases.
+
+Mid+LS can handle them.
+
+#### Story 2
+
+There are latency-insensitive tasks such as AI or stream computing, e.g. Apache Spark, that can consume a lot of resources. They require stable resources, can be suppressed and do not want to be evicted.
+
+Mid+BE can handle them.
+
+### Prerequisites
+
+Must use koordinator node reservation if someone wants to use Mid+LS
+
+### Design principles
+
+**QoS**
+
+By default, the performance of same QoS should be similar, Prod+LS and Mid+LS should be *basically* the same, also suitable to Mid+BE and Batch+BE.
+
+The adaptation of the QoS policy must be specific. In general, if the QoS is the same and the priorities are different, there is no need to make a big difference, such as the priority of eviction, and some advanced configurations in memory reclamation, usually no configuration is required.
+However, there is no finer configuration granularity for CPU group identity, so no additional customization is required for the time being.
+In the future, advanced features such as core schedule and CPU idle need to be considered.
+
+Also, you can introduce a QoS level, such as Mid between LS and BE to implement functions such as QoS detail customization.
+
+**resource oversale**
+
+Mid resource is calculated by
+```
+Allocatable[Mid] := min(Reclaimable[Mid], NodeAllocatable * thresholdRatio)
+```
+at the moment we need to change this to
+
+```
+Allocatable[Mid] := min(Reclaimable[Mid], NodeAllocatable * thresholdRatio) + Unallocated[Mid]
```

Comment:
According to my understanding, adding unallocated here is to allow Mid + LS to also use unallocated resources in the cluster, but there is a problem here.Using unallocated resources will affect the view of Prod resources, and ultimately need to support Prod's preemption of Mid, which in turn affects the stability of Mid resources.

We are also considering applying node prediction to the amplification factor of Node, so that Prod can be directly oversold, so that the Quota management and priority preemption are consistent with the native semantics of k8s.
Does this satisfy your Mid + LS need?

Reply:
Consider such a scenario,
The user deployed a Mid + LS pod, while the Mid resource of the cluster was insufficient, so the cluster autoscaler scaled out one new node.
The problem is that there are no Prod pods in the new node, so no Mid resources are available too. so we want to allowing Mid + LS to use unallocated resources in the cluster.